internal/server/middleware/hash.go

```diff
@@ -0,0 +1,100 @@
+package middleware
+
+import (
+	"bytes"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"io"
+	"net/http"
+)
+
+type HashHeader struct {
+	secret string
+}
+
+func NewHashHeader(secretKey string) *HashHeader {
+	return &HashHeader{secret: secretKey}
+}
+
+func (h *HashHeader) Validate(next http.Handler) http.Handler {
+	if h.secret == "" {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			next.ServeHTTP(w, r)
+		})
+	}
+
+	fn := func(w http.ResponseWriter, r *http.Request) {
+		requestHash := r.Header.Get("HashSHA256")
+		if requestHash == "" {
+			http.Error(w, "empty hash header", http.StatusBadRequest)
+			return
+		}
+
+		body, err := io.ReadAll(r.Body)
+		if err != nil {
+			http.Error(w, "can't read body", http.StatusBadRequest)
+			return
+		}
+
+		hash := hashHexString(body, h.secret)
+		if hash != requestHash {
+			http.Error(w, "invalid hash", http.StatusBadRequest)
+			return
+		}
+
+		r.Body = io.NopCloser(bytes.NewBuffer(body))
+		next.ServeHTTP(w, r)
+	}
+
+	return http.HandlerFunc(fn)
+}
+
+func (h *HashHeader) Sign(next http.Handler) http.Handler {
+	fn := func(w http.ResponseWriter, r *http.Request) {
+		writer := w
+		if h.secret != "" {
+			writer = newSignWriter(w, h.secret)
+		}
+
+		next.ServeHTTP(writer, r)
+	}
+
+	return http.HandlerFunc(fn)
```

Comment:
Не совсем понимаю в чем преимущество перед вот такой записью? Не совсем понятно с какой целью тут применять замыкание?
```suggestion
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		writer := w
		if h.secret != "" {
			writer = newSignWriter(w, h.secret)
		}

		next.ServeHTTP(writer, r)
	})
```

Reply:
Переделал и убрал замыкание, стало намного менее читабено! А в чем проблема с замыканием?
А ничего что next всегда передается через замыкание?